README.md
# Barcode reader widget

This repository contains a Thingworx widget allowing reading of barcodes within the browser. Both file inputs, as well as modern webcam grabbing is supported.
The underlying technology behind this widget is [QuaggaJS2](https://github.com/ericblade/quagga2).
QuaggaJS is a barcode-scanner entirely written in JavaScript supporting real- time localization and decoding of various types of barcodes such as *EAN*, *CODE 128*, *CODE 39*, *EAN 8*, *UPC-A*, *UPC-C*, *I2of5*, *2of5*, *CODE 93* and *CODABAR*. The library is also capable of using _getUserMedia_ to get direct access to the user’s camera stream. Although the code relies on heavy image-processing even recent smartphones are capable of locating and decoding barcodes in real-time.

## How to use it

By default, this is a non-reposive widget, with a text input that can be customized via CSS and a button. The wodget can also be completely hidden, and the `StartDetection` service can be used to start the detection based on external events.
The widgets offers two functioning modes, driven by the `Mode` widget property. This modes dictate what happens when clicking on the button, or when triggering the `StartDetection` service.

* Live video: a popup is automatically opened and the device camera is used to directly scan for barcodes. The camera input also includes indicators of detected codes and upon succesful detection, the popup is closed, and the input is updated to the barcode detected. The popup also offers controls for the flashlight of the device (if available).
* File input: a file dialog opens allowing the user to select a photo. Depending on the used device, this can use the camera app to take a new photo.

To understand most of the widget properties, either read the description or go to the library [website](https://serratus.github.io/quaggaJS/#configuration).

## Building and publishing

The following commands allow you to build and compile your widget:

* `npm run build`: builds the production version of the widget. Creates a new extension zip file under the `zip` folder. The production version is optimized for sharing and using in production enviroments.
* `npm run upload`: creates a build, and uploads the extension zip to the thingworx server configured in `package.json`. The build is created for developement, with source-maps enabled.
* `npm run watch`: watches the source files, and whenever they change, do a build.

CHANGELOG.md
## [1.2.1](https://github.com/ptc-iot-sharing/ThingworxBarcodeScannerWidget/compare/v1.2.0...v1.2.1) (2021-08-12)


### Bug Fixes

* Fix issue where the camera would not trigger when mode was set to File Input on iOS device. Fix [#20](https://github.com/ptc-iot-sharing/ThingworxBarcodeScannerWidget/issues/20) ([d19f369](https://github.com/ptc-iot-sharing/ThingworxBarcodeScannerWidget/commit/d19f369f86725acf87f4bcbd50b50a2425ea7684))

# [1.2.0](https://github.com/ptc-iot-sharing/ThingworxBarcodeScannerWidget/compare/v1.1.5...v1.2.0) (2021-03-12)


### Features

* Fix Migrate to using the new framework, in https://github.com/ptc-iot-sharing/ThingworxDemoWebpackWidget. Fix compatiblity with thingworx 9.1. Migrate to using https://github.com/ericblade/quagga2 ([fcf9df5](https://github.com/ptc-iot-sharing/ThingworxBarcodeScannerWidget/commit/fcf9df5806f579a13bddd95ccd1ebd0c633a52a5))


src/QuaggaJs.ide.ts
// automatically import the css file
import { ThingworxComposerWidget } from './support/widgetRuntimeSupport'

@ThingworxComposerWidget
class QuaggaJsWidget extends TWComposerWidget {

    widgetIconUrl(): string {
        return require('./images/image.svg');
    }

    widgetProperties(): TWWidgetProperties {
        require("./styles/style.css");
        return {
            'name': 'Barcode Quagga Scanner',
            'description': 'Scan barcodes and display them in mashups',
            'category': ['Common'],
            'supportsAutoResize': false,
            'properties': {
                'Width': {
                    'description': 'Total width of the widget',
                    'baseType': 'NUMBER',
                    'defaultValue': 180,
                    'isBindingTarget': false
                },
                'Height': {
                    'description': 'Total height of the widget',
                    'baseType': 'NUMBER',
                    'defaultValue': 30,
                    'isBindingTarget': false
                },
                'Code': {
                    'description': 'The code automatically detected by the scanner',
                    'baseType': 'STRING',
                    'isBindingTarget': false,
                    'isBindingSource': true
                },
                'Mode': {
                    'description': 'Input mode to use: live or image input',
                    'baseType': 'STRING',
                    'defaultValue': 'Live',
                    'selectOptions': [
                        { "text": "Live Video", "value": "Live" },
                        { "text": "File Input", "value": "Image" },
                    ]
                },
                'Barcode-Type': {
                    'description': 'Type of barcode to recognize',
                    'baseType': 'STRING',
                    "defaultValue": "Code_128",
                    'isBindingTarget': true,
                    'selectOptions': [
                        { "text": "Code 128", "value": "Code_128" },
                        { "text": "Code 39", "value": "Code_39" },
                        { "text": "Code 39 VIN", "value": "Code_39_vin" },
                        { "text": "EAN", "value": "Ean" },
                        { "text": "EAN-extended", "value": "Ean_extended" },
                        { "text": "EAN-8", "value": "Ean_8" },
                        { "text": "UPC", "value": "Upc" },
                        { "text": "UPC-E", "value": "Upc_e" },
                        { "text": "Codabar", "value": "Codabar" },
                        { "text": "ITF", "value": "I2of5" },
                        { "text": "Standard 2 of 5", "value": "S2of5" },
                        { "text": "Code 93", "value": "Code_93" }
                    ]
                },
                'Resolution': {
                    'description': 'Camera resolution to use. Depends on the target device. Applies to Live Video ONLY.',
                    'baseType': 'STRING',
                    'defaultValue': 'Re640x480',
                    'selectOptions': [
                        {"text": "320px", "value": "Re320x240"},
                        {"text": "640px", "value": "Re640x480"},
                        {"text": "800px", "value": "Re800x600"},
                        {"text": "1280px", "value": "Re1280x720"},
                        {"text": "1600px", "value": "Re1600x960"},
                        {"text": "1920px", "value": "Re1920x1080"},
                    ]
                },
                'Patch-Size': {
                    'description': 'Patch size for detecting the barcode (how big is the barcode compared with the entire image)',
                    'baseType': 'STRING',
                    'defaultValue': 'Medium',
                    'selectOptions': [
                        {"text": "x-small", "value": "X_small"},
                        {"text": "small", "value": "Small"},
                        {"text": "medium", "value": "Medium"},
                        {"text": "large", "value": "Large"},
                        {"text": "x-large", "value": "X_large"},
                    ]
                },
                'Frequency': {
                    'description': 'Detection frequency (number of scans per second). Applies to Live Video ONLY.',
                    'baseType': 'INTEGER',
                    'defaultValue': 60
                },
                'CameraFacingMode': {
                    'description': 'Specify camera to be used for the Live Video. Applies to Live Video ONLY. ',
                    'baseType': 'STRING',
                    'defaultValue': 'Environment',
                    'selectOptions': [
                        {"text": "Back-facing", "value": "Environment"},
                        {"text": "Front-facing", "value": "User"},
                    ]
                },
                'HalfSample': {
                    'description': 'Work on a half sample (half width, half height)',
                    'baseType': 'BOOLEAN',
                    'defaultValue': true
                },
                'DrawDetectionIndicator': {
                    'description': 'Enable or disable drawing of additional lines showing detection results. Applies to Live Video ONLY.',
                    'baseType': 'BOOLEAN',
                    'defaultValue': true
                }
            }
        };
    };

    widgetServices(): Dictionary<TWWidgetService> {
        return {
            StartDetection : {
                description: "Starts a new detection",
                warnIfNotBoundAsTarget: false
            }
        };
    };

    widgetEvents(): Dictionary<TWWidgetEvent> {
        return {
            'CodeDetected': {
                'warnIfNotBound': true,
                'description': 'Fired when a barcode is detected. The code variable is updated with the correct data.'
            },
            'CodeNotDetected': {
                'warnIfNotBound': true,
                'description': 'Fired when detecting a single image, but no code is found.'
            }
        };
    }

    renderHtml(): string {
        return `<div class="widget-content quagga-input-field">
                    <input class="quagga-barcode" type="text">
                    <button type="button" class="quagga-icon-barcode">&nbsp;</button>
                </div>`;
    };

    afterRender(): void {
    }

    beforeDestroy(): void {
    }

}